database/manager.go

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package database
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"go.uber.org/zap"
+)
+
+//
+// DatabaseManager writes all committed keys to the database on a timer.
+//
+
+type DatabaseManager struct {
+	logger      logging.Logger
+	keyManagers map[RelayerID]*keyManager
+	interval    time.Duration
+	db          RelayerDatabase
+}
+
+func NewDatabaseManager(logger logging.Logger, db RelayerDatabase, interval time.Duration, keys []RelayerID) *DatabaseManager {
+	keyManagers := make(map[RelayerID]*keyManager)
+	for _, key := range keys {
+		keyManager := newKeyManager(logger)
+		go keyManager.run()
+		keyManagers[key] = keyManager
+	}
+	return &DatabaseManager{
+		logger:      logger,
+		db:          db,
+		interval:    interval,
+		keyManagers: keyManagers,
+	}
+}
+
+// Run writes all committed keys to the database on the configured timer.
+// This function should only be called once.
+func (dm *DatabaseManager) Run() {
+	for range time.Tick(dm.interval) {
+		for id, km := range dm.keyManagers {
+			// Ensure we're not writing the default value
+			if km.maxCommittedHeight > 0 {
+				storedHeight, err := getLatestProcessedBlockHeight(dm.db, id)
+				if err != nil && !IsKeyNotFoundError(err) {
+					dm.logger.Error("Failed to get latest processed block height", zap.Error(err))
```

Comment:
I think an in-memory cache would be appropriate to include in the `RelayerDatabase` implementations. It would be good to verify through profiling if disk reads are even a bottleneck worth optimizing.

In any case, we need to have some kind of error check here. `IsKeyNotFoundError` is meant to filter expected versus unexpected errors.